File: p047/p47.php
<?php

/**
Distinct primes factors

The first two consecutive numbers to have two distinct prime factors are:

	14 = 2 x 7
	15 = 3 x 5

The first three consecutive numbers to have three distinct prime factors are:

	644 = 2^2 x 7 x 23
	645 = 3 x 5 x 43
	646 = 2 x 17 x 19

Find the first four consecutive integers to have four distinct prime factors.  What is the first of these numbers?
**/

require_once "../lib/prime.php";

function prime_factors_count($n) {
	return count(array_unique(prime_factors($n)));
}

$n = 2;
$expected_primes = 4;
again:
for($i = $expected_primes - 1; $i >= 0; $i--) {
	if(prime_factors_count($n + $i) != $expected_primes) {
		$n += $i + 1;
		goto again; 
	}
}

print $n . "\n";

?>

File: p047/p47.c
#include <stdio.h>
#include <stdbool.h>

#include "../lib/prime.h"

int prime_factors_count(int n) {
    prime_factors_t pf = prime_factors(n);

    int u_num = 0;
    unsigned long long unique[16] = { 0 };

    for(int i = 0; i < pf.num; i++) {
        bool found = false;
        for(int j = 0; j < u_num; j++) {
            if(unique[j] == pf.factors[i]) {
                found = true;
                break;
            }
        }

        if(!found) {
            unique[u_num] = pf.factors[i];
            u_num++;
        }
    }

    free(pf.factors);

    return u_num; // only report uniques
}

int main(int argc, char* argv[]) {

    int n = 2;
    int expected_primes = 4;

again:
    for(int i = expected_primes - 1; i >= 0; i--) {
        if(prime_factors_count(n + i) != expected_primes) {
            n += i + 1;
            goto again;
        }
    }

    printf("%i", n);

    return 0;
}

